cmd/createtree/main_test.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	// defaultTree reflects all flag defaults with the addition of a valid pk
+	defaultTree := &trillian.Tree{
+		TreeState:          trillian.TreeState_ACTIVE,
+		TreeType:           trillian.TreeType_LOG,
+		HashStrategy:       trillian.HashStrategy_RFC_6962,
+		HashAlgorithm:      sigpb.DigitallySigned_SHA256,
+		SignatureAlgorithm: sigpb.DigitallySigned_RSA,
+		DuplicatePolicy:    trillian.DuplicatePolicy_DUPLICATES_NOT_ALLOWED,
+		PrivateKey:         anyKey,
+	}
+
+	server, lis, stopFn, err := startFakeServer()
+	if err != nil {
+		t.Fatalf("Error starting fake server: %v", err)
+	}
+	defer stopFn()
+
+	validOpts := newOptsFromFlags()
+	validOpts.endpoint = lis.Addr().String()
+	validOpts.pemKeyPath = pemKey.Path
+	validOpts.pemKeyPass = pemKey.Password
+
+	nonDefaultTree := *defaultTree
+	nonDefaultTree.TreeType = trillian.TreeType_MAP
+	nonDefaultTree.SignatureAlgorithm = sigpb.DigitallySigned_ECDSA
+	nonDefaultTree.DuplicatePolicy = trillian.DuplicatePolicy_DUPLICATES_ALLOWED
+	nonDefaultTree.DisplayName = "Llamas Map"
+	nonDefaultTree.Description = "For all your digital llama needs!"
+
+	nonDefaultOpts := *validOpts
+	nonDefaultOpts.treeType = nonDefaultTree.TreeType.String()
+	nonDefaultOpts.sigAlgorithm = nonDefaultTree.SignatureAlgorithm.String()
+	nonDefaultOpts.duplicatePolicy = nonDefaultTree.DuplicatePolicy.String()
+	nonDefaultOpts.displayName = nonDefaultTree.DisplayName
+	nonDefaultOpts.description = nonDefaultTree.Description
+
+	invalidEnumOpts := *validOpts
+	invalidEnumOpts.treeType = "LLAMA!"
+
+	invalidKeyTypeOpts := *validOpts
+	invalidKeyTypeOpts.privateKeyType = "LLAMA!!"
+
+	invalidPEMPath := *validOpts
+	invalidPEMPath.pemKeyPath = "/not/a/file"
+
+	emptyPEMPass := *validOpts
+	emptyPEMPass.pemKeyPass = ""
+
+	tests := []struct {
+		desc      string
+		opts      *createOpts
+		createErr error
+		wantErr   bool
+		wantTree  *trillian.Tree
+	}{
+		{
+			desc:     "validOpts",
+			opts:     validOpts,
+			wantTree: defaultTree,
+		},
+		{
+			desc:     "nonDefaultOpts",
+			opts:     &nonDefaultOpts,
+			wantTree: &nonDefaultTree,
+		},
+		{
+			// No mandatory opts provided
+			desc:    "defaultOptsOnly",
+			opts:    newOptsFromFlags(),
+			wantErr: true,
+		},
+		{
+			desc:    "invalidEnumOpts",
+			opts:    &invalidEnumOpts,
+			wantErr: true,
+		},
+		{
+			desc:    "invalidKeyTypeOpts",
+			opts:    &invalidKeyTypeOpts,
+			wantErr: true,
+		},
+		{
+			desc:    "invalidPEMPath",
+			opts:    &invalidPEMPath,
+			wantErr: true,
+		},
+		{
+			desc:    "emptyPEMPass",
+			opts:    &emptyPEMPass,
+			wantErr: true,
+		},
+		{
+			desc:      "createErr",
+			opts:      validOpts,
+			createErr: errors.New("create tree failed"),
+			wantErr:   true,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		func() {
+			var out *string
+			println = func(a ...interface{}) (int, error) {
+				str := fmt.Sprintln(a...)
+				out = &str
+				return 0, nil
+			}
+			defer func() {
+				println = fmt.Println
+			}()
```

Comment:
I've refactored it so that `createTree` takes `fmt.Println` as an argument. The test should be more straightforward as well (although it looks largely the same, there are no defer or println global anymore).